This lab on Logistic Regression is a Python adaptation of p. 161-163 of "Introduction to Statistical Learning with Applications in R" by Gareth James, Daniela Witten, Trevor Hastie and Robert Tibshirani. Adapted by R. Jordan Crouser at Smith College for SDS293: Machine Learning (Spring 2016).

In [1]:
import pandas as pd
import numpy as np

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis as QDA
from sklearn.metrics import confusion_matrix, classification_report, precision_score

%matplotlib inline

# 4.6.3 Linear Discriminant Analysis

Let's return to the ${\tt Smarket}$ data from ${\tt ISLR}$. 

In [3]:
df = pd.read_csv('data/Smarket.csv', usecols=range(1,10), index_col=0, parse_dates=True)
# df = pd.read_csv('data/Smarket.csv')
df.head(15)

Lag1   Lag2   Lag3   Lag4   Lag5  Volume  Today Direction
Year                                                                  
2001-01-01  0.381 -0.192 -2.624 -1.055  5.010  1.1913  0.959        Up
2001-01-01  0.959  0.381 -0.192 -2.624 -1.055  1.2965  1.032        Up
2001-01-01  1.032  0.959  0.381 -0.192 -2.624  1.4112 -0.623      Down
2001-01-01 -0.623  1.032  0.959  0.381 -0.192  1.2760  0.614        Up
2001-01-01  0.614 -0.623  1.032  0.959  0.381  1.2057  0.213        Up
2001-01-01  0.213  0.614 -0.623  1.032  0.959  1.3491  1.392        Up
2001-01-01  1.392  0.213  0.614 -0.623  1.032  1.4450 -0.403      Down
2001-01-01 -0.403  1.392  0.213  0.614 -0.623  1.4078  0.027        Up
2001-01-01  0.027 -0.403  1.392  0.213  0.614  1.1640  1.303        Up
2001-01-01  1.303  0.027 -0.403  1.392  0.213  1.2326  0.287        Up
2001-01-01  0.287  1.303  0.027 -0.403  1.392  1.3090 -0.498      Down
2001-01-01 -0.498  0.287  1.303  0.027 -0.403  1.2580 -0.189      Down
2001-01-01 -0.189 -0.498  0.287  1.303  0.027  1.0980  0.680        Up
2001-01-01  0.680 -0.189 -0.498  0.287  1.303  1.0531  0.701        Up
2001-01-01  0.701  0.680 -0.189 -0.498  0.287  1.1498 -0.562      Down

Now we will perform LDA on the ${\tt Smarket}$ data from the ${\tt ISLR}$ package. In ${\tt Python}$, we can fit a LDA model using the ${\tt LDA()}$ function, which is part of the ${\tt lda}$ module of the ${\tt sklearn}$ library. As we did with logistic regression and KNN, we'll fit the model using only the observations before 2005, and then test the model on the data from 2005.

In [3]:
X_train = df[:'2004'][['Lag1','Lag2']]
y_train = df[:'2004']['Direction']

X_test = df['2005':][['Lag1','Lag2']]
y_test = df['2005':]['Direction']

lda = LDA()
model = lda.fit(X_train, y_train)

print(model.priors_)

[0.49198397 0.50801603]


The LDA output indicates prior probabilities of ${\hat{\pi}}_1 = 0.492$ and ${\hat{\pi}}_2 = 0.508$; in other words,
49.2% of the training observations correspond to days during which the
market went down.

In [4]:
print(model.means_)

[[ 0.04279022  0.03389409]
 [-0.03954635 -0.03132544]]


The above provides the group means; these are the average
of each predictor within each class, and are used by LDA as estimates
of $\mu_k$. These suggest that there is a tendency for the previous 2 days’
returns to be negative on days when the market increases, and a tendency
for the previous days’ returns to be positive on days when the market
declines. 

In [5]:
print(model.coef_)

[[-0.05544078 -0.0443452 ]]


The coefficients of linear discriminants output provides the linear
combination of ${\tt Lag1}$ and ${\tt Lag2}$ that are used to form the LDA decision rule. In other words, these are the multipliers of the elements of ${X = x}$ in 4.19:

${\delta_k(x) = x^T\Sigma^{-1}\mu_k - \frac{1}{2}\mu_k^T\Sigma^{-1}\mu_k + log\pi_k}$

If $−0.0554\times{\tt Lag1}−0.0443\times{\tt Lag2}$ is large, then the LDA classifier will
predict a market increase, and if it is small, then the LDA classifier will
predict a market decline. **Note**: these coefficients differ from those produced by ${\tt R}$.



The ${\tt predict()}$ function returns a list of LDA’s predictions about the movement of the market on the test data:

In [6]:
pred=model.predict(X_test)
print(np.unique(pred, return_counts=True))

(array(['Down', 'Up'], dtype='<U4'), array([ 70, 182]))


In [7]:
from IPython.display import Image

The model assigned 70 observations to the "Down" class, and 182 observations to the "Up" class. Let's check out the confusion matrix to see how this model is doing. We'll want to compare the **predicted class** (which we can find in ${\tt pred}$) to the **true class** (found in ${\tt y\_test})$.

![image.jpg](table4-6.jpg)

![image.jpg](table4-7.jpg)

In [8]:
print(confusion_matrix(pred, y_test))
print(classification_report(y_test, pred, digits=3))

[[ 35  35]
 [ 76 106]]
              precision    recall  f1-score   support

        Down      0.500     0.315     0.387       111
          Up      0.582     0.752     0.656       141

    accuracy                          0.560       252
   macro avg      0.541     0.534     0.522       252
weighted avg      0.546     0.560     0.538       252



In [9]:
classification_report?

We can also get the predicted _probabilities_ using the ${\tt predict\_proba()}$ function:

In [10]:
pred_p = model.predict_proba(X_test)

Applying a 50% threshold to the posterior probabilities allows us to recreate
the predictions:

In [11]:
print(np.unique(pred_p[:,1]>0.5, return_counts=True))

(array([False,  True]), array([ 70, 182]))


Notice that the posterior probability output by the model corresponds to
the probability that the market will **increase** (opposite of R output in text):

In [12]:
print(np.stack((pred_p[10:20,1], pred[10:20])).T)

[['0.5093037238790318' 'Up']
 ['0.4880011537380811' 'Down']
 ['0.510484773063352' 'Up']
 ['0.5293238777881214' 'Up']
 ['0.5255407143881711' 'Up']
 ['0.5200416608518921' 'Up']
 ['0.5064224705341396' 'Up']
 ['0.4969106228816935' 'Down']
 ['0.5021193878585957' 'Up']
 ['0.5113669134834818' 'Up']]


If we wanted to use a posterior probability threshold other than 50% in
order to make predictions, then we could easily do so. For instance, suppose
that we wish to predict a market decrease only if we are very certain that the
market will indeed decrease on that day—say, if the posterior probability
is at least 90%:

In [13]:
print(np.unique(pred_p[:,1]>0.9, return_counts=True))

(array([False]), array([252]))


No days in 2005 meet that threshold! In fact, the greatest posterior probability
of decrease in all of 2005 was 54.2%:

In [14]:
max(pred_p[:,1])

0.5422132554518978

# 4.6.4 Quadratic Discriminant Analysis
We will now fit a QDA model to the ${\tt Smarket}$ data. QDA is implemented
in ${\tt sklearn}$ using the ${\tt QDA()}$ function, which is part of the ${\tt qda}$ module. The
syntax is identical to that of ${\tt LDA()}$.

In [16]:
qda = QDA()
model2 = qda.fit(X_train, y_train)
print(model2.priors_)
print(model2.means_)

[0.49198397 0.50801603]
[[ 0.04279022  0.03389409]
 [-0.03954635 -0.03132544]]


The output contains the group means. But it does not contain the coefficients
of the linear discriminants, because the QDA classifier involves a
_quadratic_, rather than a linear, function of the predictors. The ${\tt predict()}$
function works in exactly the same fashion as for LDA.

In [17]:
pred2=model2.predict(X_test)
print(np.unique(pred2, return_counts=True))
print(confusion_matrix(pred2, y_test))
print(classification_report(y_test, pred2, digits=3))

(array(['Down', 'Up'], dtype=object), array([ 50, 202]))
[[ 30  20]
 [ 81 121]]
              precision    recall  f1-score   support

        Down      0.600     0.270     0.373       111
          Up      0.599     0.858     0.706       141

    accuracy                          0.599       252
   macro avg      0.600     0.564     0.539       252
weighted avg      0.599     0.599     0.559       252



Interestingly, the QDA predictions are accurate almost 60% of the time,
even though the 2005 data was not used to fit the model. This level of accuracy
is quite impressive for stock market data, which is known to be quite
hard to model accurately. 

This suggests that the quadratic form assumed
by QDA may capture the true relationship more accurately than the linear
forms assumed by LDA and logistic regression. However, we recommend
evaluating this method’s performance on a larger test set before betting
that this approach will consistently beat the market!

# An Application to Carseats Data
Let's see how the ${\tt LDA/QDA}$ approach performs on the ${\tt Carseats}$ data set, which is
included with ${\tt ISLR}$. 

Recall: this is a simulated data set containing sales of child car seats at 400 different stores.

In [93]:
df2 = pd.read_csv('data/Carseats.csv')
df2.head(10)
# df2.shape

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   
5  10.81        124     113           13         501     72       Bad   78   
6   6.63        115     105            0          45    108    Medium   71   
7  11.85        136      81           15         425    120      Good   67   
8   6.54        132     110            0         108    124    Medium   76   
9   4.69        132     113            0         131    124    Medium   76   

   Education Urban   US  
0         17   Yes  Yes  
1         10   Yes  Yes  
2         12   Yes  Yes  
3         14   Yes  Yes  
4         13   Yes   No  
5         16    No  Yes  
6         15   Yes   No  
7         10   Yes  Yes  
8         10    No   No  
9         17    No  Yes

See if you can build a model that predicts ${\tt ShelveLoc}$, the shelf location (Bad, Good, or Medium) of the product at each store. Don't forget to hold out some of the data for testing!

In [85]:
# Your code here
predictors = ['Sales', 'CompPrice', 'Price']
train_len = round(len(df2) * 0.70)
X_train = df2[:train_len][predictors]
X_test = df2[train_len:][predictors]
y_train = df2[0:train_len]["ShelveLoc"]
y_test = df2[train_len:]["ShelveLoc"]

In [86]:
model2 = lda.fit(X_train, y_train)
print(model2.priors_)

[0.24285714 0.18928571 0.56785714]


In [87]:
print(model2.coef_)

[[-0.85070287  0.08097587 -0.08174572]
 [ 1.34498179 -0.13053429  0.13018438]
 [-0.08450466  0.00888024 -0.00843436]]


In [88]:
pred2 = model.predict(X_test)

In [89]:
print(confusion_matrix(pred2, y_test))

[[13  0  3]
 [ 0 17  8]
 [15 15 49]]


In [90]:
print(classification_report(y_test, pred2, digits=3))

              precision    recall  f1-score   support

         Bad      0.812     0.464     0.591        28
        Good      0.680     0.531     0.596        32
      Medium      0.620     0.817     0.705        60

    accuracy                          0.658       120
   macro avg      0.704     0.604     0.631       120
weighted avg      0.681     0.658     0.649       120



To get credit for this lab, please post your answers to the following questions:

- What was your approach to building the model?
- How did your model perform?
- Was anything easier or more challenging than you anticipated?

to Piazza: https://piazza.com/class/igwiv4w3ctb6rg?cid=23